jep/302/README.adoc

```diff
@@ -0,0 +1,260 @@
+
+== Reasoning
+
+=== Why Git
+
+Using filesystem-level tools offering a snapshotting feature, like LVM, ZFS or btrfs to give a few examples, was considered.
+But this was evacuated because _Essentials_ vision is about providing an link:https://github.com/jenkinsci/jep/tree/71d9391744c8cc7d6595805f7fdd327eedf6811a/jep/300#automatically-updated-distribution["_easier to use_ and _easier to manage_ Jenkins environment"].
+As per the link:https://github.com/jenkinsci/jep/tree/71d9391744c8cc7d6595805f7fdd327eedf6811a/jep/300#target-audience[targeted audience], we obviously do not want to expect _Essentials_ users to be system experts able to set up a dedicated filesystem to operate Jenkins.
+And even with system expert, doing so would not make Essentials a very easy and quick to use distribution of Jenkins.
+
+Git offers in this matter a powerful user-space tool that allows us to version,
+and quickly rollback to some previous state if need be.
+
+Git is also a very common tool nowadays for developers,
+hence it will help making Essentials more accessible to contributors.
+
+== Why not use compatibleSinceVersion metadata
+
+TIP: A given plugin can indicate what is "the oldest version [...] configuration-compatible with.".
+In other words, if I am changing the configuration format of a plugin in non backwards compatible manner, I am _expected_ to express it as such using link:https://jenkinsci.github.io/maven-hpi-plugin/hpi-mojo.html#compatibleSinceVersion[`compatibleSinceVersion`].
+
+So, this means we could use this field to _optimize_ the way we roll back: once a configuration file has been upgraded by a plugin update, we can exclude it from rollback if the `compatibleSinceVersion` field was not changed.
+
+We decided to not do, for now at least, for two reasons:
+
+* this looks like an _optimization_. $
+Hence as such, this is probably premature to try and be very smart with the way the downgrade will work ;
+* there is no guarantee that all plugin developers currently know and think about updating this metadata when needed.
+This is likely something that will be improveable when we work on link:https://issues.jenkins-ci.org/browse/JENKINS-49806[JENKINS-49806] and write the associated Jenkins Enhancement Proposal.
+
+== Backwards Compatibility
+
+There are no backwards compatibility concerns related to this proposal.
+
+== Security
+
+=== Secrets
+
+Versioning secrets is not an issue, as the snapshotting system is going to be local to the running instance.
+As the Git repository data will never be pushed _outside_, no data leak is expected from this side.
+
+=== Man In The Middle
+
+The main issue here is that an attacker could for instance instruct the evergreen client to ignore everything (by putting `*` in `.gitignore`), hence make it impossible to rollback.
```

Comment:
don;t use .gitignore then.  use `-f` to add?
move builds outside JENKINS_HOME  (have a directly at the mount that is /builds /whatever /config  and make "config JENKINS_HOME where the usual stuff is stored).

Reply:
> don;t use .gitignore then. use -f to add?

Not convinced this fondamentally changes the issue. The goal is still to be able to upgrade the evergreen-client too. So, be it by hacking the `.gitignore` content, or by hacking what things are passed to `-f` I guess the issue is to really find robust ways to not accept instructions from some wrong _instructors_.

 > move builds outside JENKINS_HOME (have a directly at the mount that is /builds /whatever /config and make "config JENKINS_HOME where the usual stuff is stored).

Good point. I gave it some thoughts, but kept it under JENKINS_HOME to _respect_ JEP-301:

>  [it is expected that all mutable state and data will be written and stored in JENKINS_HOME, including: Data generated by the Jenkins process. ...](https://github.com/jenkinsci/jep/tree/master/jep/301#mutable-data)

Reply:
just because that is written does not mean it is correct - infact that secrtion of JEP is bad bad bad...

why? 
--pluginRoot --webroot.   JENKINS_HOME is possibly mounted on a resiliant storage (and is usually slower)  there is no need to unpack the jenkins plugins or war into jenkins-home...

Also logs... Hmmm if they are in jenkins home and you rollback then you loose some logs..
that;s not a good thing is it?

now if jep-301 wanted to say thal all data be contained under a single directory then fine (but JENKINS_HOME does not need to be at the root of that directory) IMO and if you are making a more complex system because of that descision then go revisit it...